### PR TITLE
Display playback progress during clip sharing

### PR DESCRIPTION
**Pull request by MiSikora**
## Description

This PR adds playback progress to clip sharing. Playback progress is draggable. There are also some changes to the behavior.

* Play/pause button now pauses playback instead of stopping it.
* Changing start or end of a clip stops playback.

Note: Zooming is acting junky when timeline is not scrolled to the beginning. I plan to work on it in the future.

## Testing Instructions

1. Start sharing.
2. Drag progress to some position.
3. Play the clip.
4. Playback should start from the dragged position.
5. Let the clip playback to the end.
6. Play the clip again.
7. Playback should start.
8. Pause playback.
9. Playback should stop.
10. Resume playback.
11. Playback should start from the paused position.
12. While clip is playing drag end or start of the clip.
13. Playback should stop.
14. Resume playback.
15. Playback should start from the beginning.

## Screenshots or Screencast 

https://github.com/Automattic/pocket-casts-android/assets/30936061/20a73f2d-d12b-452e-aa12-798030bd9434

## Checklist
- [x] ~If this is a user-facing change, I have added an entry in CHANGELOG.md~
- [x] Ensure the linter passes (`./gradlew spotlessApply` to automatically apply formatting/linting)
- [x] I have considered whether it makes sense to add tests for my changes
- [x] ~All strings that need to be localized are in `modules/services/localization/src/main/res/values/strings.xml`~
- [x] Any jetpack compose components I added or changed are covered by compose previews
- [x] ~I have updated (or requested that someone edit) [the spreadsheet](https://docs.google.com/spreadsheets/d/107jqrutZhU0fVZJ19SBqxxVKbV2NWSdQC9MFYdLiAxc/edit?usp=sharing) to reflect any new or changed analytics.~
 
#### I have tested any UI changes...

- [x] ~with different themes~
- [x] ~with a landscape orientation~
- [x] ~with the device set to have a large display and font size~
- [x] ~for accessibility with TalkBack~

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>ShareClipViewModelListener</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by CookieyedCodes** [comment]
Ummmmm would it be possible to have a download clip function and speed control for those downloads

**Comment by MiSikora** [comment]
This isn't something that we're currently planning for the feature.

**Comment by CookieyedCodes** [comment]
Fair enough, I get why but ye can understand why I would love both tbh 😅

File: modules/features/clip/src/main/kotlin/au/com/shiftyjelly/pocketcasts/clip/ClipPlayer.kt [hunk 1]
```diff
@@ -50,18 +71,47 @@ interface ClipPlayer {
...
+    private val coroutineScope = CoroutineScope(Dispatchers.Main.immediate)
+
     override val isPlayingState = MutableStateFlow(false)
     override val errors = MutableSharedFlow<Exception>()
+    override val playbackProgress = channelFlow {
+        while (currentCoroutineContext().isActive) {
+            // Instead of simple delay loop we use inner job that we join.
+            // This is done to account for dynamic changes to scale and resolution of the timeline.
+            // When the timeline is zoomed in we want to send events much more frequently to
+            // make the playback progress smooth.
```
Comment (ashiagr) [inline modules/features/clip/src/main/kotlin/au/com/shiftyjelly/pocketcasts/clip/ClipPlayer.kt:98]: Thanks for adding this comment 👍

**Review by ashiagr — APPROVED** [review]
Works well in my testing 👍

**Pull request merged**